### DISCUSSION TITLE
SnippetViewSet.list_display: Allow flexibility in column type and position for 'action' column (??)

### DISCUSSION THREAD

**Discussion by enzedonline** [wagtail/wagtail#12171] [Ideas]
Not sure if 'action' column is the right term - the column with the link and action buttons.

Currently, setting anything other than a string for the first element of  `list_display` renders the viewset without any link or action buttons on the items. This is down to [this logic](https://github.com/wagtail/wagtail/blob/54e108744bc8fa556ca2368f742f635874b5ad04/wagtail/admin/views/generic/models.py#L315) in the generic view. 

It would be much more useful to allow a custom column type to be the 'action' column. My understanding is that this was a design decision to handle the case that not all column types would be compatible. If, for some reason, the custom column type is not compatible, then an error can be thrown here since it would always be caught in development before production.

The 'action' column is also hard-coded to be the first column in the list. It could be useful in some circumstances to designate a different column as the 'action' column. For example:

```python
    list_display = [ImageColumn("image"), "title", "get_department_subcategory", "sku", UpdatedAtColumn()]
```

In this case, it would make more sense to make `"title"` the action column, but visually more sense to display the image in the first column.

**Comment by SebCorbin** [comment]
The code your reference is the solution, `_get_title_column` only ensures there's a `TitleColumn` by default in the first position. But you can set it yourself, that would result in 
```python
    list_display = [ImageColumn("image"), TitleColumn("title"), "get_department_subcategory", "sku", UpdatedAtColumn()]

**Comment by enzedonline** [comment]
Unfortunately not. There is no action column with the following:
```python
list_display = [ImageColumn("image"), TitleColumn("title"), "get_department_subcategory", "sku", UpdatedAtColumn()]
```
Even if you have `TitleColumn` for your first column, the edit link and action menu is missing. 
```python
list_display = [TitleColumn("title"), "get_department_subcategory", "sku", UpdatedAtColumn()]
```
The `ImageColumn` in the example is subclassed from `TitleColumn` in any case.

I spoke with Sage regarding this in the Slack channel a while back, he mentioned it was a deliberate choice at the time to handle the cases where the column type might not be compatible with the edit link and action menu.
